pkg/controllers/clusterresourcesnapshot/watcher.go

```diff
@@ -0,0 +1,105 @@
+	snapshotKRef := klog.KRef(name.Namespace, name.Name)
+
+	startTime := time.Now()
+	klog.V(2).InfoS("Reconciliation starts", "clusterResourceSnapshot", snapshotKRef)
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("Reconciliation ends", "clusterResourceSnapshot", snapshotKRef, "latency", latency)
+	}()
+
+	if err := r.Client.Get(ctx, name, &snapshot); err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(4).InfoS("Ignoring NotFound clusterResourceSnapshot", "clusterResourceSnapshot", snapshotKRef)
+			return ctrl.Result{}, nil
+		}
+		klog.ErrorS(err, "Failed to get clusterResourceSnapshot", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, controller.NewAPIServerError(true, err)
+	}
+	// check if it is the first resource snapshot
+	_, exist := snapshot.Annotations[fleetv1beta1.NumberOfResourceSnapshotsAnnotation]
+	if !exist {
+		// we only care about when a new resource snapshot index is created
+		klog.V(2).InfoS("Ignore the subsequent sub resource snapshots", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, nil
+	}
+	// check if it is the latest resource snapshot
+	isLatest, err := strconv.ParseBool(snapshot.Labels[fleetv1beta1.IsLatestSnapshotLabel])
+	if err != nil {
+		err := fmt.Errorf("invalid annotation value %s", fleetv1beta1.IsLatestSnapshotLabel)
+		klog.ErrorS(err, "resource snapshot has does not have valid is latest annotation", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, controller.NewUnexpectedBehaviorError(err)
+	}
+	if !isLatest {
+		// All newly created resource snapshots should start with the latest label to be true.
+		// However, this can happen if the label is removed fast by the time this reconcile loop is triggered.
+		klog.V(2).InfoS("Newly created resource snapshot %s is not the latest", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, nil
+	}
+	// get the CRP name from the label
+	crp := snapshot.Labels[fleetv1beta1.CRPTrackingLabel]
+	if len(crp) == 0 {
+		err := fmt.Errorf("invalid CRPTrackingLabel label value %s", crp)
+		klog.ErrorS(err, "Invalid clusterResourceSnapshot", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, controller.NewUnexpectedBehaviorError(err)
+	}
+	// enqueue the CRP to the rollout controller queue
+	r.RolloutController.Enqueue(crp)
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Watcher) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&fleetv1beta1.ClusterResourceSnapshot{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		WithEventFilter(predicate.Funcs{
+			// skipping delete events as rollout controller doesn't handle that.
+			DeleteFunc: func(_ event.DeleteEvent) bool {
+				return false
+			},
+			// skipping update events as the specs are immutable.
+			UpdateFunc: func(evt event.UpdateEvent) bool {
```

Comment:
Hi Ryan! The spec is surely immutable, but the metadata is still subject to change; we probably shouldn't ignore these events.

Reply:
which metadata change should we handle?